debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java

```diff
@@ -1512,6 +1512,50 @@ else if (additionalCondition.isPresent()) {
         return sql.toString();
     }
 
+    public String buildUnionBasedSelectWithRowLimits(TableId tableId, int limit, String projection, List<String> boundConditions,
+                                                     Optional<String> additionalCondition, List<String> orderByColumnNames) {
+        final String TABLE_ALIAS = "t";
+        final String outerOrderBy = orderByColumnNames.stream()
+                .map(c -> quotedColumnIdString(TABLE_ALIAS) + "." + quotedColumnIdString(c))
+                .collect(Collectors.joining(", "));
+        final String innerOrderBy = orderByColumnNames.stream()
+                .map(this::quotedColumnIdString)
+                .collect(Collectors.joining(", "));
+        final String outerQueryPrefix = "SELECT * FROM (";
+        final String outerQuerySuffix = " ORDER BY " + outerOrderBy + " LIMIT " + limit;
+
+        StringBuilder queryBuilder = new StringBuilder(outerQueryPrefix);
+
+        if (boundConditions.size() > 1) {
+            // If there are multiple bound conditions exist (pk), use the complex UNION query
+            queryBuilder.append(boundConditions.stream()
+                    .map(condition -> buildSelectWithRowLimits(
+                            tableId,
+                            limit,
+                            projection,
+                            Optional.of(condition),
+                            additionalCondition,
+                            innerOrderBy))
+                    .collect(Collectors.joining(") UNION (", "(", ")")));
```

Comment:
I think we should prefer "UNION ALL". I expect it to be more efficient since it doesn't have to look for duplicates (and we can't have duplicates here (and if we do, they should not be eradicated)).

Reply:
Query conditions are like:
```
(k1 > <k1_min>) AND (k1 <= <k1_max>)
(k1 = <k1_min> AND k2 > <k2_min>) AND (k1 <> <k1_max> OR k2 <= <k2_max>)
...
```

The first condition `(k1 > <k1_min>) AND (k1 <= <k1_max>)` and the second part of the WHERE condition `(k1 <> <k1_max> OR k2 <= <k2_max>)` may create duplicates.